### DISCUSSION TITLE
State of SEO on Solidus (Meta Discussion)

### DISCUSSION THREAD

**Discussion by fthobe** [solidusio/solidus#5905] [New Features or Ideas]
I have two fears in life: bad SEO and @jarednorman banning me from writing solidus issues for life so upon popular request this is a discussion :) 

schema.org and the google desired reference [implementation](https://developers.google.com/search/docs/appearance/structured-data/intro-structured-data) became over the years an important factor on search rankings the most notable case being free google shopping ads upon implementation.

Before I write anymore issues for the bucket list:

- is what's considered SEO best practice is considered desirable for the core? 
- is what's considered SEO best practice desirable for the starter template? 

In my personal opinion a baseline of fields required by schema.org and google shopping should at least be out of the box possible and it's low hanging fruit: 

- 5-7 fields on products 
- a couple of fields on taxons
- schema.org for the front end

**Comment by kennyadsl** [comment]
Not sure about its state, but this has been the go-to solution for adding some SEO enhancement in Solidus: https://github.com/karmakatahdin/solidus_seo. It's probably worth refreshing the extension and it should be good to go for what's needed here.

**Reply by freelancedigitalmarketer** [reply]
State of SEO on Solidus -[ Freelance Digital Marketer in Wayanad](https://suriyas.com/)

Solidus is an open-source e-commerce platform that offers flexibility and customization, enabling businesses to create powerful, scalable, and feature-rich online stores. SEO (Search Engine Optimization) on Solidus requires thoughtful strategies that align with the platform's capabilities, as well as a clear focus on optimizing content and website structure for search engines. If you're a freelance digital marketer in Wayanad, here’s how SEO can be effectively implemented on Solidus to achieve great results:

1. On-Page SEO
Keyword Research: Begin by identifying the most relevant and high-traffic keywords for your niche. As a freelance digital marketer in Wayanad, you should aim for a balance of local and industry-specific keywords. Use tools like Google Keyword Planner, Ahrefs, or SEMrush to uncover high-value keywords such as "freelance digital marketing Wayanad" or "SEO services Wayanad."
Title Tags & Meta Descriptions: Ensure that each page on your Solidus store has unique and descriptive title tags and meta descriptions. For instance, for a landing page, use a title like "Freelance Digital Marketer in Wayanad | Professional SEO Services" and an engaging description.
Content Optimization: Optimize your content with targeted keywords, without keyword stuffing. Include the keyword "freelance digital marketer in Wayanad" in places like headings, product descriptions, and blog posts.
Mobile Optimization: Solidus is mobile-responsive, but ensure that your content, images, and buttons are all optimized for mobile-first indexing, given that most local searches are conducted via mobile devices.
2. Technical SEO
Site Speed: Solidus allows for extensive customization, but one area to focus on is ensuring fast page load times. Tools like Google PageSpeed Insights can provide suggestions on how to improve performance. Compress images, minify CSS and JavaScript, and leverage browser caching.
Structured Data (Schema Markup): Add structured data to your product pages, blog posts, and contact pages. Use schema types such as LocalBusiness and Service to help search engines understand your services better. For example, a LocalBusiness schema could include your name, address, and phone number (NAP), which helps improve local SEO.
XML Sitemap & Robots.txt: Ensure that your Solidus store has a properly configured XML sitemap and that you’re using a robots.txt file to control which pages should be indexed by search engines.
3. Local SEO for Wayanad
Google My Business: As a freelancer based in Wayanad, it's crucial to set up and optimize your Google My Business (GMB) profile. Make sure your NAP information is consistent across your site and local directories.
Local Listings and Citations: Build local citations by getting listed on relevant local directories and platforms. This strengthens your presence in the Wayanad area.
Location-Specific Keywords: Use location-specific terms in your content, like "freelance digital marketer in Wayanad" or "SEO services for businesses in Wayanad," to help attract local search traffic.
Local Reviews: Encourage clients to leave reviews on platforms such as Google, Facebook, and local business directories. Positive reviews not only build credibility but also improve your local search rankings.
4. Content Marketing & Link Building
Blog and Resource Creation: Create helpful content such as blog posts, guides, and case studies that target long-tail keywords. As a freelancer, you could write about topics such as "How to Improve Your Website's SEO in Wayanad" or "Top Digital Marketing Strategies for Small Businesses in Wayanad."
Guest Posting & Link Building: Build high-quality backlinks to your Solidus site through guest posting, collaborations, or writing for local Wayanad-based businesses or media outlets. Quality backlinks are still a critical ranking factor.
Social Media Marketing: Social signals play an indirect role in SEO. Build a strong presence on platforms like Facebook, Instagram, and LinkedIn, which can also help generate traffic to your Solidus site.
5. Tracking & Analytics
Google Analytics & Search Console: Implement Google Analytics and Google Search Console to track website performance. You can track metrics like traffic, bounce rate, and keywords ranking, and use these insights to optimize your Solidus site further.
Conversion Tracking: Set up conversion tracking on your site (such as contact form submissions or newsletter sign-ups) to ensure you're capturing valuable data for your digital marketing efforts.
Conclusion:
As a freelance digital marketer in Wayanad, implementing effective SEO strategies on a Solidus-based e-commerce site requires attention to both technical SEO and content marketing. By focusing on local SEO, optimizing on-page elements, improving site speed, and creating valuable content, you can improve your online visibility and attract more clients in the Wayanad area. Keep up with SEO best practices, continuously analyze your data, and adjust your strategies for optimal performance.

**Reply by jarednorman** [reply]
Haha, I won't ban you @fthobe! I will however ban @freelancedigitalmarketer for posting LLM-generated garbage.

**Comment by fthobe** [comment]
# Edits
Fixed two references with correct names, fixed some typos, added 301 redirects and URL structure in what's missing

# Basic technical Information
Currently best practice would be to transport structured data via json, it's the most wildly accepted format. 
There are alternatives, but there some problems: 

- google does officially support json via jscript injection, others don't (support varies largely here), json is the by Google recommended format (which indicates that itemprop will go EOL). 
- itemprop has good results on most pages but seen the above I wouldn't use it, none of the search engines do officially support itemprop injection
- RDFa I list only as a courtesy, never seen it, never used it, I hardly know it and it hasn't seen a large adoption

# What fields do we have to play with
- [X] description fields on taxons and products do accept and render HTML correctly
- [X] we do have prices for products and variants
- [X] we have product names 
- [X] titles 
- [X] meta description

# What's missing: 
- [x] brands ( @jarednorman suggested in #4621 `Spree::Product#brand` which has the great aspect that it's guaranteed backwards compatibility, but it opens doors to wrong configuration), #5989 PR made
- [x] Price Ranges especially on category pages to model aggregatedoffer > it will be deprecated for SEO in favor of option groups and [structured data carousels](https://developers.google.com/search/docs/appearance/structured-data/carousels-beta) (right now every does a personal flavour implementation of [itemcarousel](https://developers.google.com/search/docs/appearance/structured-data/carousel) mostly abusing the carousel (currently supported for Course list, Movie Recipe, Restaurant) which is while not documented mostly gracefully accepted by Google for offers (though not documented), but I think it has an immensely strong case on Front-End usability. Given the varying success of json injection I wouldn't want to do that via jscript.
- [x] Reviews: 1-5 stars scale with endpoints to return a count of reviews and an average value, given the varying success of json injection I wouldn't want to do that via jscript.
- [ ] Variant URLs: Google expects for google shopping that a variant has an URL that upon page load is preselected (this can be done in front-end) and rendering a canonical link to the parent item, an approach that validates is a link to the product with the variant sku as parameter (eg BigCommerce does that). The best case would be a slug per variant, but that is hard to maintain from a content perspective (it's a lot of work to create tons of variants with different content such as description, meta title,...). A middle ground would be: one slug per variant with a canonical link on the primary product. 
- [ ] 301 Redirects 
- [ ] Discussion about shortening slug moving "/products/" and "/t/categories/" to root 

# What's cool and nice to have: 
- date published 
- date updated > crawling rhythm will be impacted

# What should be deprecated
Nothing right now

# What should be deprecated
- Excessive usage of heading tags
- itemprops (as things are going it will be deprecated by Google)

# What could be deprecated for Europe and US but might be needed on other markets
- [ ] html keywords in head section: Quoting here "Meta keywords are largely outdated for major search engines like Google, Yahoo, and Bing, but some engines, such as Yandex and Baidu, may still factor them into their ranking algorithms" and this is current best practice. 

# What could be achieved afterwards relatively easily purely with front-end out of the box with the starter front-end
- Full Google Analytics compliance through having the information available for events
- Full Structured Data Compliance through structured data json
- Full Google Merchant Center Compliance (Google Shopping without Feed) through structured data json
- Facebook Catalogue will work

# What will mostly work
- Bing / Yahoo is working on implementing variant groups, the issue here is that all countries work with the same underlying bing version so success could vary, they are usually trailing in the us by 12-18 months and in Europe by up to 36 months;
- Yandex works relatively reliable with some shorter or longer delay according to Google's standards, newer and experimental stuff like variantitemlists might not work yet, pages will work relatively well;
- Baidoo I have too little experience and the documentation is only in Chinese and requires login, back then when I worked in China we had good results just following european practices, but we surely left some stones unturned. 

# What other features have shown good / great results
- a sitemap should be there, it became universal standard and I would go with any sitemap is better than no sitemap, shooting for the moon I would also do an HTML sitemap (we will surely do it), an example is the [booking.com html sitemap](https://www.booking.com/region.en-gb.html). 
- a xml sitemap
- content control for the home page including opening hours and regions

**Reply by tvdeyen** [reply]
> # What could be achieved afterwards relatively easily purely with front-end out of the box with the starter front-end
> * Full Google Analytics compliance through having the information available for events
> * Full Structured Data Compliance through structured data json
> * Full Google Merchant Center Compliance (Google Shopping without Feed) through structured data json
> * Facebook Catalogue will work

I am all in for semantic markup and structured data, but will never be comfortable with baking in custom platform related code into core and/or starter frontend. I like the idea of making it an installer option, but all custom platform related code should live in extensions.

Again, the code we generate or provide should always be as semantic, well structured and therefore machine readable as possible, with customization options for tracking extensions if necessary. Any good tracking technology will be able to read well structured code anyway.

Everything else has already been said much better by Jared and Alberto.

**Reply by fthobe** [reply]
ok, to better understand your point: reviews for you are custom, correct?

**Reply by tvdeyen** [reply]
> ok, to better understand your point: reviews for you are custom, correct?

you mean https://github.com/solidusio-contrib/solidus_reviews ?

**Reply by fthobe** [reply]
yes, in my oppinion reviews became generic and baseline regarding SEO to a point that Google will drop your schema if none of reviews / offers are available, it has also been sufficiently standartized on a 1-5 scale to a point that even facebook standartizes it's own Thumbs up / Thumbs down to a 5 star scale. 

I think reviews have sufficiently matured to backport it core to have full structured data support. 

To be blunt, I wouldn't have something that basic to store without significant business logic and that is a basic requirement of what's google for ecommerce now depend on additional maintenance.

**Reply by tvdeyen** [reply]
> yes, in my oppinion reviews became generic and baseline regarding SEO to a point that Google will drop your schema if none of reviews / offers are available, it has also been sufficiently standartized on a 1-5 scale to a point that even facebook standartizes it's own Thumbs up / Thumbs down to a 5 star scale.
> 
> I think reviews have sufficiently matured to backport it core to have full structured data support.
> 
> To be blunt, I wouldn't have something that basic to store without significant business logic and that is a basic requirement of what's google for ecommerce now depend on additional maintenance.

If we agree on removing Google from this equation I tend to agree.

**Reply by fthobe** [reply]
So your point is structured data yes, google analytics no, is that correct and would that mean you support the idea of integrating reviews to core?

Also there's the discussion open regarding the [integration of GA4](https://github.com/solidusio/solidus/discussions/4621) where it seemed to agreeable to implement GA4 in the starter frontend.

**Reply by tvdeyen** [reply]
> So your point is structured data yes, google analytics no, is that correct and would that mean you support the idea of integrating reviews to core?

structured data yes, google analytics ~~no~~ related code in an extension.

> you support the idea of integrating reviews to core?

I think this makes sense. Maybe we (@solidusio/core-team) could start thinking about feature flags instead of fine grained extensions for basic features like this. Maintaining so many extensions is a lot for a small team like ours. And unmaintained extensions do not shed a lot of good light on our ecosystem.

> Also there's the discussion open regarding the [integration of GA4](https://github.com/solidusio/solidus/discussions/4621) where it seemed to agreeable to implement GA4 in the starter frontend.

From @jarednorman's [comment](https://github.com/solidusio/solidus/discussions/4621#discussioncomment-11191817):

> I would like the starter frontend to ship with this as an option.

I assume he means that this should be an installer option. If not I will ban **him** 😆

**Reply by fthobe** [reply]
> I assume he means that this should be an installer option. If not I will ban him 😆

Hey, we all have been there wanting to ban Jared. 

But on this rare occasion I would like to cheer him: 
It makes sense to build out the starter template to what it is, a reference implementation (also for us to prepare a custom template later on), while I share your flavour of thought on everything google maybe it's wise if the starter-frontend is this reference to include what's the industry standard, despite me and you sharing the same sentiment here. 

Regarding the burden of maintenance
Honestly we haven't found one extension that works trouble free right now, I can see how multi domain is an edge case, just review and brands taxons are not (it would be in my opinion a very difficult case to make to say otherwise) and in my opinion having subscriptions easily running from the start has a very strong business case as it has the potential to dominate a (not that small) nieche with the current offering on the market especially considering that the barrier between products and services become ever more fluent. 

To return to GA and looking at the various default templates out there on the various platforms GA and structured data is a feature and not technological debt and it belongs in my opinion in the starter template bucket. #banmeaswell

**Reply by jarednorman** [reply]
@tvdeyen is correct. I indeed meant that it would be an option.

Reviews are something that the vast majority of our stores depend on third parties for. It's not a candidate for inclusion in core in my mind, but I would love if the extension were fixed up to provide good support for the starter frontend.

**Reply by fthobe** [reply]
What is the mechanism to verify the presence of an extension or the activation of a feature? I can see how we can easily override behavior with a config file, what mechanism do you have in mind to verify extensions available?

**Reply by jarednorman** [reply]
The Solidus installer already does this for things like payment methods. You can take a look at how that works.

**Comment by jarednorman** [comment]
@fthobe I don't want to sidetrack the conversation, so I'm posting this outside the thread, but you've used the word "depreciate" and "depreciated" throughout your comments above. You mean "deprecate" and "deprecated". English is a terrible language.

**Reply by fthobe** [reply]
@jarednorman Sorry Jared, I know you are right, I have to converse in 5 languages and macOS / iOS does not handle that very graceful. Please accept my apologies, I will be more careful in the future. I actually do profoundly love English. 
![IMG_4603](https://github.com/user-attachments/assets/de529b03-c1c7-47fd-bbcd-b5bf0a00d9c5)

**Reply by jarednorman** [reply]
No need for apologizing! It's a very common mistake that native English speakers make too. I wouldn't have even pointed it out except that the word appears in the code itself, so I thought it would be helpful for you to know the correct one.

**Comment by fthobe** [comment]
# Fields needed 

| Component                                                                                                                    | Field(s) Required                                                                                                                                                                                                                                                                                                                   |
|------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Spree Core Product [Merchant Listing](https://developers.google.com/search/docs/appearance/structured-data/merchant-listing) | product sku per variant<br>\*product GTIN (EAN o similar text string)per variant<br>\*condition (per variant as defined by schema.org) [New, Refurbished, Used, Damaged]\*<br>\*breadcrumb_taxon (only one taxon)<br>\*slug per variant (in addition to current product)<br>image(s)<br>product name (ideally also variant name but we can fake through variant option combinations)<br>is variant of parent item<br>brand name<br>brand page url<br>product description<br>product price and currency code (ISO)<br>availability  |
| Spree Core Store [Organization](https://developers.google.com/search/docs/appearance/structured-data/organization)           | \*Legal Name (text string)<br>\*taxID (text string)<br>\*Address (break down on multiple lines as in google structured data)<br>Logo<br>\*Image (nice to have)<br>\*Contact Phone (text string)<br>\*Contact Email (text string)<br>\*VAT ID (text string)<br>\*description                                                                                                                                                                                                                                         |
| solidus_reviews                                                                                                              | aggregate review including review count and average review on a 1-5 scale<br>individual reviews with 1-5 rating and comment                                                                                                                                                                                                            |

# PRs to be created
PR 1: GTIN and Condition
PR 2: Variant Slug and Breadcrumb_Taxon
PR 3: All Organization Details

# Notes
## Breadcrumbs
Discussion regarding breadcrumbs: https://github.com/solidusio/solidus_starter_frontend/issues/130
Copy mechanism for regular taxon and write to DB as regular taxon, but just belongs to ONE taxon instead of many.
